### PR TITLE
N°824 - Allow OQL JOIN on object ON EventNotification

### PR DESCRIPTION
**Pull request by accognet**
Add a new object_class field in EventNotification to be used in a future version of iTop

### CODE REVIEW

**Comment by Molkobain** [comment]
Support PRs review:
- Anne-Cath will ask Vincent if we want to display the "Notifications" tab even when no trigger but existing notifications (code to be done)
- Data migration during setup has been done (mind to reset "previous version " to 3.2 instead of 3.3)

File: datamodels/2.x/itop-structure/module.itop-structure.php [hunk 1]
```diff
@@ -228,6 +228,54 @@ public static function AfterDatabaseCreation(Config $oConfiguration, $sPreviousV
 					SetupLog::Info("... default triggers/action successfully created for $iClassesWithLogCount classes.");
 				}
 			}
+
+			//N°824 - Fill object_class in EventNotification from the Triggers target_class
+			if (version_compare($sPreviousVersion, '3.3.0', '<')) {
```
Comment (Molkobain) [inline datamodels/2.x/itop-structure/module.itop-structure.php]: Mind to restore to `3.2.0`

File: datamodels/2.x/itop-structure/module.itop-structure.php [hunk 2]
```diff
@@ -228,6 +228,54 @@ public static function AfterDatabaseCreation(Config $oConfiguration, $sPreviousV
...
+
+				$oSearch = DBObjectSearch::FromOQL('SELECT TriggerOnObject');
+				$oSet = new DBObjectSet($oSearch);
+				$aData = [];
+				while ($oTrigger = $oSet->Fetch()) {
+					$aData[$oTrigger->GetKey()] =  $oTrigger->Get('target_class');
+				}
+
+				foreach ($aData as $sKey => $sTargetClass) {
+
```
Comment (Molkobain) [inline datamodels/2.x/itop-structure/module.itop-structure.php:431]:
```suggestion
```

File: datamodels/2.x/itop-structure/module.itop-structure.php [hunk 3]
```diff
@@ -228,6 +228,54 @@ public static function AfterDatabaseCreation(Config $oConfiguration, $sPreviousV
...
+						//in this case, we have toget the name of the final class
+						$sTableToRead = MetaModel::DBGetTable($sTargetClass, 'finalclass');
+						$oAttDefToRead = MetaModel::GetAttributeDef($sTargetClass, 'finalclass');
+						$aColumnsToReads = array_keys($oAttDefToRead->GetSQLColumns());
+						$sColumnToRead = $aColumnsToReads[0]; // We know that a string has only one column
+						$sObjectPrimaryKey = MetaModel::DBGetKey($sTargetClass);
+
+						$sRepair = "UPDATE `$sTableToSet` JOIN `$sTableToRead` ON `$sTableToSet`.`$sColumnObjectId` = `$sTableToRead`.`$sObjectPrimaryKey` SET `$sTableToSet`.`$sColumnToSet` = `$sTableToRead`.`$sColumnToRead` WHERE  `$sTableToSet`.`$sColumnTriggerId` = '".$sKey."' AND `$sTableToSet`.`$sColumnToSet` = ''";
+					} else {
+
```
Comment (Molkobain) [inline datamodels/2.x/itop-structure/module.itop-structure.php:442]:
```suggestion
```

**Comment by rquetiez** [comment]
Data migration during might have an impact on the duration of the setup. This impact has to be evaluated and documented.

**Comment by Molkobain** [comment]
Functional review: Accepted for 3.3.

There are still tests to do on a huge database (1.5M+ EventNotifications) to anticipate queries duration during the setup; and we are too close to the 3.2 code freeze to manage it.
Tests have already be done on a smaller DB (300k EventNotifications), results can be found on the drive under _R&D > Studies_

**Comment by Molkobain** [comment]
@accognet please : 
  * Change target branch to develop
  * "Rebase onto" your branch on develop instead of support/3.2
  * Resolve conflicts
  * Then you can squash & merge on develop

**Comment by Hipska** [comment]
Hi, any chance this bug fix also gets applied to 3.2 since it is a LTS?

**Comment by Molkobain** [comment]
Hello Thomas, as this was a non blocking issue and as it was a bit risky it was decided back then not to apply it on the `support/3.2` branch.

Following your comment I rediscussed it with the POs and the decision remains the same.

**Comment by Hipska** [comment]
I think the text, legends and so on need to be adapted here: https://www.itophub.io/wiki/page?id=3_3_0:release:start

(It currently gives the impression bugs will get fixed in LTS versions)

**Comment by Molkobain** [comment]
Which parts exactly?

Bugs are fixed in the LTS of course (77 for iTop 3.2.3), but not necessary all of them.
If we consider one is too risky regarding what it brings, it will only be fixed in the next major version like this bug.

**Pull request merged**